### PR TITLE
fix(PSGO-37): stop OrderedMap guessing []string from array contents

### PR DESCRIPTION
**Pull request by Matovidlo**
Jira: https://linear.app/keboola/issue/PSGO-37/investigate-keboolago-utils-change

**Changes:**
- `pkg/orderedmap/json.go`: delete `convertStringSliceIfPossible`/`convertNestedSlices` and their 2 call sites in the decoder. Arrays now always decode to `[]any`, matching `encoding/json`'s stdlib behavior for `map[string]interface{}` (this is what `OrderedMap` did before v1.4.1/PR #60).
- `pkg/orderedmap/orderedmap.go`: add an exported, opt-in `ToStringSlice(value any) ([]string, error)` helper for callers that know a specific field's schema is a string array.
- `pkg/orderedmap/json_test.go`: rewrite the 2 tests that encoded the old coercion as a requirement; add a regression test for the empty-array case and a table test for `ToStringSlice`.

**Why:** v1.4.1 (PR #60) made the JSON decoder silently retype *any* array to `[]string` whenever every element happened to be a string at runtime - including every empty array, unconditionally, regardless of the field's true schema. That's value-dependent type coercion, not schema-based, and it broke every existing `.([]any)` assertion in downstream consumers wherever a field is always-string content or sometimes empty - concretely in keboola-as-code: orchestrator `dependsOn` (phase links silently dropped), `shared_code_row_ids` (every shared-code-linking transformation errors), JSON-Schema `required`/`allOf`/enum processing (silent no-ops), default-bucket rewriting and transformation block loading (skipped on empty arrays). Two attempts to adopt v1.4.1 in keboola-as-code were reverted same-day because of this (`eacc44026`, `bd8a26178`).

A prior attempt at a fix (#68) went in this same direction - disabling the coercion - but left the functions commented out rather than removed, and was never merged. This PR completes that direction properly: clean removal, explicit opt-in replacement, and tests that guard the specific regressions found (empty array, all-string array).

The field name used in the test (`rowsSortOrder`) is the one referenced by the originating bug report, but `Config.RowsSortOrder` in keboola-sdk-go is actually a plain typed `[]string` struct field decoded directly by jsoniter's reflection decoder - it never passes through `OrderedMap` at all. So this global coercion could not have fixed its own stated motivation even in principle. The real call site that needs `ToStringSlice` is being identified in the keboola-as-code adoption PR (follow-up).

Deliberately split from the pure Go-version bump (PSGO-38, #71) so the two can be reviewed and rolled back independently.

Verified: `go build ./...`, `go vet ./...`, `task tests` (64 tests, 1 pre-existing unrelated skip), `golangci-lint run` with the CI-pinned v2.0.2 binary - all clean.

-----------

### CODE REVIEW

**Comment by Matovidlo** [comment]
@claude review

File: pkg/orderedmap/orderedmap.go [hunk 1]
```diff
@@ -389,6 +389,29 @@ func visit(key Path, valueRaw any, parent any, callback VisitCallback) {
...
+	out := make([]string, len(arr))
+	for i, item := range arr {
+		s, ok := item.(string)
+		if !ok {
+			return nil, fmt.Errorf(`expected string at index %d, found "%T"`, i, item)
+		}
+		out[i] = s
+	}
+	return out, nil
+}
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply pkg/orderedmap/orderedmap.go]: **Fixed** ✅ in commit 1aa69a7f0affb37f3df2a5536c0a27023072705a

Good catch. ToStringSlice now switches on the value's type: []string is returned as-is, []any is validated/converted element-by-element, anything else errors. So a value set programmatically via .Set() with a native []string, or one decoded from JSON/YAML ([]any), both work.

File: pkg/orderedmap/json_test.go [hunk 2]
```diff
@@ -363,32 +363,77 @@ func TestOrderedMap_UnmarshalJSON_Struct(t *testing.T) {
...
+
+	ss, err = ToStringSlice([]any{"a", "b"})
+	assert.NoError(t, err)
+	assert.Equal(t, []string{"a", "b"}, ss)
+
+	_, err = ToStringSlice("not a slice")
+	assert.ErrorContains(t, err, `expected []any, found "string"`)
+
+	_, err = ToStringSlice([]any{"a", 1})
+	assert.ErrorContains(t, err, `expected string at index 1, found "int"`)
```
[the remark this answers is not in the document]
Comment (Matovidlo) [reply pkg/orderedmap/json_test.go]: **Fixed** ✅ in commit 1aa69a7f0affb37f3df2a5536c0a27023072705a

Added a case in TestToStringSlice covering ToStringSlice([]string{"a","b"}) and updated the wrong-type error assertion to the new wording: `expected []any or []string, found "string"`.

File: pkg/orderedmap/orderedmap.go [hunk 3]
```diff
@@ -389,6 +389,33 @@ func visit(key Path, valueRaw any, parent any, callback VisitCallback) {
 	}
 }
 
+// ToStringSlice converts a value obtained from OrderedMap (e.g. via Get or GetNested) into a
```
Comment (pepamartinec) [inline pkg/orderedmap/orderedmap.go]: Only exported symbol in this region — wedged between the unexported `visit` and `convertToMap`. Move it next to the other exported API, or into its own file (`convert.go`).

File: pkg/orderedmap/orderedmap.go [hunk 4]
```diff
@@ -389,6 +389,33 @@ func visit(key Path, valueRaw any, parent any, callback VisitCallback) {
...
+// []string. OrderedMap itself never guesses a field's schema from its runtime content - JSON/YAML
+// arrays always decode to []any - so callers that know a specific field is a string array must
+// opt in explicitly via this helper. Accepts either []any (the JSON/YAML-decoded shape) or
+// []string directly (e.g. a value set programmatically via Set). A nil value converts to an
+// empty, non-nil []string.
+func ToStringSlice(value any) ([]string, error) {
+	switch v := value.(type) {
+	case nil:
+		return []string{}, nil
```
Comment (pepamartinec) [inline pkg/orderedmap/orderedmap.go]: Returns empty + no error, so a missing field is indistinguishable from `[]`. With `GetNestedOrNil`, a typo'd path silently yields `[]string{}` — the same silent-no-op this PR fixes downstream (`shared_code_row_ids`, JSON-Schema `required`). Also asymmetric with `["a", null]`, which errors at index 1.

If the lenient behaviour is intended, the doc comment has to say callers check `Get`/`GetNested`'s `found` first. It currently states the behaviour without the guidance.

File: pkg/orderedmap/orderedmap.go [hunk 5]
```diff
@@ -389,6 +389,33 @@ func visit(key Path, valueRaw any, parent any, callback VisitCallback) {
...
+	case []string:
+		return v, nil
```
Comment (pepamartinec) [inline pkg/orderedmap/orderedmap.go]: Aliases the input; the `[]any` branch below copies. Sorting or writing in place to the result mutates the value inside the `OrderedMap` in this case only, though the two inputs are presented as interchangeable. `slices` is already imported:

```suggestion
	case []string:
		return slices.Clone(v), nil
```

File: pkg/orderedmap/json_test.go [hunk 6]
```diff
@@ -363,32 +363,82 @@ func TestOrderedMap_UnmarshalJSON_Struct(t *testing.T) {
...
 	m := New()
 	assert.NoError(t, json.Unmarshal([]byte(in), &m))
 	v, _ := m.Get("arr")
 	outer, ok := v.([]any)
 	assert.True(t, ok)
-	inner0, ok := outer[0].([]string)
-	assert.True(t, ok)
-	assert.Equal(t, []string{"a", "b"}, inner0)
-	inner1, ok := outer[1].([]string)
+	inner0, ok := outer[0].([]any)
```
Comment (pepamartinec) [inline pkg/orderedmap/json_test.go:393]: `assert.True(t, ok)` on 395 does not stop the test, so a failed assertion panics here on `outer[0]` (index out of range, nil slice) instead of failing with the message. Same at 380 → 381 and 415 → 416. `require.True` for the type assertions, `assert` for the value checks.

File: pkg/orderedmap/json_test.go [hunk 7]
```diff
@@ -363,32 +363,82 @@ func TestOrderedMap_UnmarshalJSON_Struct(t *testing.T) {
...
+	assert.NoError(t, json.Unmarshal([]byte(in), &m))
+	v, ok := m.Get("arr")
 	assert.True(t, ok)
-	assert.Equal(t, []string{"c"}, inner1)
+	arr, ok := v.([]any)
+	assert.True(t, ok, "expected []any, got %T", v)
+	assert.Empty(t, arr)
+}
+
+func TestToStringSlice(t *testing.T) {
```
Comment (pepamartinec) [inline pkg/orderedmap/json_test.go]: Not a table test, despite the description — seven sequential `assert` pairs sharing `ss`/`err`. Failures report a bare line number, and later cases run against a stale `ss`. Table + `t.Run` for the documented entry point of the new API.

File: pkg/orderedmap/json_test.go [hunk 8]
```diff
@@ -363,32 +363,82 @@ func TestOrderedMap_UnmarshalJSON_Struct(t *testing.T) {
...
 	assert.True(t, ok)
-	_, isAny := v.([]any)
-	assert.False(t, isAny)
-	ss, ok := v.([]string)
-	assert.True(t, ok)
+	arr, ok := v.([]any)
+	assert.True(t, ok, "expected []any, got %T", v)
+	assert.Equal(t, []any{"a", "b", "c"}, arr)
+
+	ss, err := ToStringSlice(v)
```
Comment (pepamartinec) [inline pkg/orderedmap/json_test.go]: Already covered by `TestToStringSlice`. This test's contract is the decoder shape; drop these three lines.

File: pkg/orderedmap/json_test.go [hunk 9]
```diff
@@ -363,32 +363,82 @@ func TestOrderedMap_UnmarshalJSON_Struct(t *testing.T) {
...
+	in := `{"arr":[]}`
+	m := New()
+	assert.NoError(t, json.Unmarshal([]byte(in), &m))
+	v, ok := m.Get("arr")
 	assert.True(t, ok)
-	assert.Equal(t, []string{"c"}, inner1)
+	arr, ok := v.([]any)
+	assert.True(t, ok, "expected []any, got %T", v)
+	assert.Empty(t, arr)
+}
```
Comment (pepamartinec) [inline pkg/orderedmap/json_test.go:414]: All three tests assert the type only. The v1.4.1 breakage was behavioural and inside this package: `GetNested("dependsOn[0]")` and `SetNested("dependsOn[1]", "z")` returned `expected array found "[]string"`, and `VisitAllRecursive` never descended into a string array — `orderedmap.go:166/204/288/383` type-switch on `[]any`.

Add one test that decodes `{"dependsOn":["a","b"]}` and exercises those three. It documents why the type matters and blocks the next attempt to "fix" `rowsSortOrder` this way.

**Review by pepamartinec — CHANGES_REQUESTED** [review]
Clean revert of the json.go half of #60, and correct. Problems:

**1. `go.mod` has no `retract` for v1.4.1.** It is still the latest tag, so `go get -u github.com/keboola/go-utils` resolves to the version this PR exists to undo.

```go
retract v1.4.1 // JSON decoder retyped string arrays to []string, breaking GetNested/SetNested/VisitAllRecursive.
```

**2. Target tag is unstated.** This adds an exported symbol, so it is v1.5.0, not v1.4.2. Name it in the description so the keboola-as-code adoption PR pins the right version.

Rest inline.

File: pkg/orderedmap/orderedmap.go (hunk 3 above)
Comment (Matovidlo) [reply pkg/orderedmap/orderedmap.go]: **Fixed** ✅ in commit 48268f7b66571b889827d75bcff54e33f0232fec

Moved ToStringSlice into its own file, pkg/orderedmap/convert.go, next to the package's other conversion-flavored code rather than wedged between visit and convertToMap.

File: pkg/orderedmap/orderedmap.go (hunk 4 above)
Comment (Matovidlo) [reply pkg/orderedmap/orderedmap.go]: **Fixed** ✅ in commit 48268f7b66571b889827d75bcff54e33f0232fec

Agreed - the lenient path was wrong, not just under-documented. ToStringSlice(nil) now returns an error instead of []string{}, so a typo'd/missing path via GetNestedOrNil surfaces instead of silently producing an empty slice. Added a case to the new table-driven TestToStringSlice covering this explicitly. Checked the one real caller (keboola-as-code's TypeStringArray.ParseValue) - it passes a typed nil []string in the empty-string case, which still hits the []string branch, not this one, so this doesn't change behavior there for that path. Flagged separately: ParseValue can reach ToStringSlice with a true untyped nil if called with nil directly, which will now error where it previously didn't - tracking as a follow-up on the keboola-as-code side.

File: pkg/orderedmap/orderedmap.go (hunk 5 above)
Comment (Matovidlo) [reply pkg/orderedmap/orderedmap.go]: **Fixed** ✅ in commit 48268f7b66571b889827d75bcff54e33f0232fec

Applied your suggestion verbatim - slices.Clone(v) in the []string branch. Added TestToStringSlice_DoesNotAliasStringInput as a direct regression test: sets a []string via Set, converts it, mutates the result, and asserts the value stored in the OrderedMap is unaffected.

File: pkg/orderedmap/json_test.go (hunk 6 above)
Comment (Matovidlo) [reply pkg/orderedmap/json_test.go:393]: **Fixed** ✅ in commit 48268f7b66571b889827d75bcff54e33f0232fec

Changed all three type-assertion gates (the ones whose failure would panic on the next line via a nil-slice index, not just fail a value comparison) to require.True, keeping assert for the actual value checks (Equal/Empty) that follow.

File: pkg/orderedmap/json_test.go (hunk 7 above)
Comment (Matovidlo) [reply pkg/orderedmap/json_test.go]: **Fixed** ✅ in commit 48268f7b66571b889827d75bcff54e33f0232fec

Rewrote TestToStringSlice as a table test with named cases run via t.Run, including the new nil-is-now-an-error case. Each case gets its own ss/err scope, so nothing carries over stale state, and a failure reports the case name.

File: pkg/orderedmap/json_test.go (hunk 8 above)
Comment (Matovidlo) [reply pkg/orderedmap/json_test.go]: **Fixed** ✅ in commit 48268f7b66571b889827d75bcff54e33f0232fec

Dropped the ToStringSlice call and its two assertions from TestOrderedMap_UnmarshalJSON_ArrayOfStrings_StaysAny - that test now only asserts the decoder produces []any, which is its actual contract.

File: pkg/orderedmap/json_test.go (hunk 9 above)
Comment (Matovidlo) [reply pkg/orderedmap/json_test.go:414]: **Fixed** ✅ in commit 48268f7b66571b889827d75bcff54e33f0232fec

Added TestOrderedMap_DecodedStringArray_PathAccessors, decoding {"dependsOn":["a","b"]} and exercising exactly the three internal paths you named: GetNested("dependsOn[0]"), SetNested("dependsOn[1]", "z") followed by GetNested to confirm the write landed, and VisitAllRecursive confirming it descends into both indices. Verified this test fails with "expected array found \"[]string\"" if I temporarily reintroduce the coercion, confirming it actually exercises the SliceStep/[]any type-switch path you pointed at (orderedmap.go's GetNestedPath/SetNestedPath and visit).

File: pkg/orderedmap/convert.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+	"slices"
+)
+
+// ToStringSlice converts a value obtained from OrderedMap (e.g. via Get or GetNested) into a
+// []string. OrderedMap itself never guesses a field's schema from its runtime content - JSON/YAML
+// arrays always decode to []any - so callers that know a specific field is a string array must
+// opt in explicitly via this helper. Accepts either []any (the JSON/YAML-decoded shape) or
+// []string directly (e.g. a value set programmatically via Set).
+//
+// A nil value is an error, not an empty slice: nil is what GetNestedOrNil returns for a missing
```
Comment (pepamartinec) [inline pkg/orderedmap/convert.go:14]: This overclaims — only *untyped* nil errors. Verified against 48268f7:

```
ToStringSlice(nil)           -> error `expected []any or []string, found "<nil>"`
ToStringSlice([]string(nil)) -> (nil, nil)          // slices.Clone preserves nil
ToStringSlice([]any(nil))    -> ([]string{}, nil)
```

Three flavours of nil, three outcomes, and the table covers only the first. A JSON `null` decodes to an untyped nil so the `GetNestedOrNil` case you are guarding does error — but anyone reading this comment and passing a nil-valued `[]string` from `Set` gets `nil, nil` silently.

Either narrow the wording to untyped nil, or normalise: return `[]string{}` for the typed-nil cases so the helper never hands back a nil slice with a nil error. Either way, add table cases for `[]string(nil)` and `[]any(nil)` — the current three-way split is untested and will drift.

File: pkg/orderedmap/json_test.go [hunk 11]
```diff
@@ -363,32 +364,131 @@ func TestOrderedMap_UnmarshalJSON_Struct(t *testing.T) {
...
+
+	var visited []string
+	m.VisitAllRecursive(func(path Path, _ any, _ any) {
+		visited = append(visited, path.String())
+	})
+	assert.Contains(t, visited, "dependsOn[0]")
+	assert.Contains(t, visited, "dependsOn[1]")
+}
+
+func TestToStringSlice(t *testing.T) {
```
Comment (pepamartinec) [inline pkg/orderedmap/json_test.go:447]: `ToStringSlice` moved to `convert.go`, but its two tests stayed here. Discoverability was the reason for the move; `convert_test.go` for `TestToStringSlice` and `TestToStringSlice_DoesNotAliasStringInput`.

**Review by pepamartinec — COMMENTED** [review]
Round 2. All 7 inline points are genuinely fixed — I replicated the two claims that could have been wishful:

- Reintroducing the coercion (`git checkout origin/main -- pkg/orderedmap/json.go`) makes `TestOrderedMap_DecodedStringArray_PathAccessors` fail with `path "dependsOn": expected array found "[]string"`, plus the three `StaysAny` tests. The test exercises the path you claimed.
- Dropping `slices.Clone` makes `TestToStringSlice_DoesNotAliasStringInput` fail with `actual: []string{"mutated", "b"}`.

`go vet`, `go test ./pkg/orderedmap/... ./pkg/deepcopy/... -race`, and `golangci-lint run -c build/ci/golangci.yml` are clean on 48268f7. Remaining:

**1. `retract v1.4.1` still missing** — round 1, unaddressed and unanswered. `go.mod` has no `retract`, v1.4.1 is still the latest tag, so `go get -u github.com/keboola/go-utils` still resolves to the version this PR undoes.

```go
retract v1.4.1 // JSON decoder retyped string arrays to []string, breaking GetNested/SetNested/VisitAllRecursive.
```

If you are deliberately skipping it, say why and I will drop it.

**2. Target tag still unstated** — round 1, unaddressed. New exported symbol ⇒ v1.5.0, not v1.4.2. Name it in the description so the keboola-as-code adoption PR pins the right version.

**3. Description is now stale on three counts:**
- "`pkg/orderedmap/orderedmap.go`: add an exported, opt-in `ToStringSlice`" — it is in `convert.go` as of 48268f7.
- "`golangci-lint run` with the CI-pinned v2.0.2 binary" — `scripts/tools.sh` pins v2.9.0 since #71 merged.
- `nil` being an error is the helper's API contract and was decided in review; it belongs in the description, not only in the commit message.

The nil edge is inline.

**Review by pepamartinec — APPROVED** [review]
Claude jeste nasel par veci, ale prijde mi uz spis nit style, takze klidne bych to mergnul
